example/scripts/sample_csv_data_loader.py

```diff
@@ -26,155 +10,15 @@
 if len(sys.argv) > 2:
     neo_host = sys.argv[2]
 
-es = Elasticsearch([
-    {'host': es_host if es_host else 'localhost'},
-])
-
 NEO4J_ENDPOINT = 'bolt://{}:7687'.format(neo_host if neo_host else 'localhost')
-neo4j_endpoint = NEO4J_ENDPOINT
+ES_ENDPOINT = es_host if es_host else 'localhost'
 
-neo4j_user = 'neo4j'
-neo4j_password = 'test'
+csv_job_factory = CsvToNeo4jJobFactory(NEO4J_ENDPOINT, 'neo4j', 'test')
+es_job_factory = Neo4jToEsJobFactory(ES_ENDPOINT, NEO4J_ENDPOINT, 'neo4j', 'test')
```

Comment:
I think the sample loader is just an example for quick start and we should make it easy to extend with new metadata. Make it a factory will make it coupled. And what happen if we add dashboard extractor etc? I think the databuilder is designed to work with different source and sync and having the factory would be duplicate a lot functionality which has already existed.

And I think for ETL, we should still use some actual DAG / job like https://github.com/lyft/amundsendatabuilder/blob/master/example/dags/sample_dag.py.

Reply:
I'm not sure I follow. The factories can actually reduce duplication, as they can be used to create jobs easier. Also, we don't necessarily need to use the factories in the sample loader, I mostly put it there, so we can see it used. 

> And what happen if we add dashboard extractor etc?

Then we could it directly, or, build a factory for it. 

> And I think for ETL, we should still use some actual DAG / job like https://github.com/lyft/amundsendatabuilder/blob/master/example/dags/sample_dag.py.

I think the problem, with that DAG is that it is incomplete. It does not do everything that is needed for a full load, such as elastic search loading, last update loading, etc. If we added all of that, without the factories, it would be significantly more repeated code. 

Please take a look at my new version of sample_csv_data_loader and note how much is reduced by the factory code. And, that reduced code is now usable from other jobs, dags and classes quite easily. I've left the original sample_data_loader in place, for comparison, it's still quite large, and a bit hard to parse IMHO, maybe compare them?

Ultimately, this is what I'm working towards, it's my DAG I'm building out for our csv based solution:
```
from datetime import datetime, timedelta
from airflow.models import DAG
from airflow.operators import DummyOperator
from airflow.operators.python_operator import PythonOperator
from airflow.utils.dates import days_ago
import boto3

from databuilder.job.factory.neo4j_to_es_job_factory import Neo4jToEsJobFactory
from databuilder.job.factory.csv_to_neo4j_job_factory import CsvToNeo4jJobFactory

#TODO - parameterize env
NEO4J_ENDPOINT = 'bolt://amundsen-neo4j-dev:7687'
ES_ENDPOINT = 'amundsen-elasticsearch-dev'

csv_job_factory = CsvToNeo4jJobFactory(NEO4J_ENDPOINT, 'neo4j', 'test')
es_job_factory = Neo4jToEsJobFactory(ES_ENDPOINT, NEO4J_ENDPOINT, 'neo4j', 'test')

args = {
  'owner': 'selliott',
  'start_date': days_ago(1),
  'depends_on_past': False,
}

def download_from_s3(bucket_name, key):
    s3 = boto3.resource('s3')
    f_name = key.split("/")[-1]
    full_file_name = "/tmp/{}".format(f_name)

    print("downloading [{}] from [{}] to [{}] on worker pod".format(key, bucket_name, full_file_name))

    s3.meta.client.download_file(bucket_name, key, full_file_name)
    
def run_csv_job(file_loc, table_name, model):
    csv_job_factory.create_model_job(file_loc, table_name, model).launch()    

dag = DAG(dag_id='load_amundsen_dag', schedule_interval='@once', default_args=args)

start_task = DummyOperator(task_id='start_task', dag=dag)

#TODO - parameterize bucket name
download_tables_csv_s3_task = PythonOperator(
  task_id='download_tables_csv_s3_task',
  python_callable=download_from_s3,
  op_kwargs={
    'bucket_name': 'bucket-name',
    'key': 'amundsen/meta_tables/tables.csv'
  },
  dag=dag)

load_tables_amundsen_task = PythonOperator(
  task_id='load_tables_amundsen_task',
  python_callable=run_csv_job,
  op_kwargs={
    'file_loc': '/tmp/tables.csv',
    'table_name': 'run_csv_job',
    'model': 'databuilder.models.table_metadata.TableMetadata'
  },
  dag=dag)

start_task >> download_tables_csv_s3_task >> load_tables_amundsen_task
```

If I don't have the factories in place, then, all of that code that is in the sample data loader must be repeated in my DAG, which is quite a lot. 

Is there a better way? What do you think?

Reply:
@feng-tao I've thought about this some more. I think it comes down to, if anyone else in the community would find this useful, either now, or at some future point. I personally think it could be useful to others, but, I could be wrong. Perhaps if we don't agree, we could get another opinion like @jornh or another major contributor? If after that, it still appears that it might not be useful to others, I can close the PR and just move it to an internal release. wdyt?

If there is anything I can change, that would make this better, please don't hesitate to lmk.

Reply:
@javamonkey79 , I understand it helps to reduce duplication in this scenario, but what we aim is to make databuilder modularize for executor, transformer, loader, publisher and allow "lego" like plug and play for different metadata which could be table stats, table watermark, table last updated, table usage, table owner, table cost, table source, table lineage , user info, dashboard owner, etc(list keeps going). Each metadata could need different extractor, loader, publisher etc. The factory fixes with csv_extractor, and Fs loader, neo4j publisher which is hard to extend compared to the existing method. And it also takes away the flexibility to allow user to specify configuration.

At Lyft, our table databuilder job has ~20 airflow tasks each of which handles one metadata ingestion with different extractor(rs, lyftpg, hive, presto, etc). It would be hard to extend with the factory method compared to https://github.com/lyft/amundsendatabuilder/blob/master/example/dags/sample_dag.py which you could just add a new task for each new metadata.

Reply:
> what we aim is to make databuilder modularize for executor, transformer, loader, publisher and allow "lego" like plug and play for different metadata such as table stats, table watermark, table last updated, table usage, table owner, table cost, table source, table lineage , user info, dashboard owner, etc(list keeps going). Each metadata could need different extractor, loader, publisher etc. 

Yes, I see that architecture, and it's good. Providing a factory does not detract from it, in fact, it extends it. The great thing about the approach I have provided, is that users could still use the bare bones ETL flow, my aim is to ease the setup of all of the instructions needed to setup a job, to reduce the boiler plate and make it simpler. I think other such factory classes could be created around other metadata sources and databases as well, to further simplify ETL work loads. 

> The factory fixes with csv_extractor, and Fs loader, neo4j publisher which is hard to extend compared to the existing method. And it also takes away the flexibility to allow user to specify configuration.

Yes, that's true, that the config is baked in. I have considered adding a config object to the constructor args of each factory, such that it takes precedence over the default factory configuration. If I added this, would it help?

> At Lyft, our table databuilder job has ~20 airflow tasks each of which handles one metadata ingestion with different extractor(rs, lyftpg, hive, presto, etc). It would be hard to extend with the factory method compared to https://github.com/lyft/amundsendatabuilder/blob/master/example/dags/sample_dag.py which you could just add a new task for each new metadata.

I am interested to see how you have done it, compared to how we are doing it. It's a bit difficult to understand how there is not quite a lot of duplication in your DAG code, without seeing it first hand.

Reply:
I will be OOO rest of the end the week. Will provide some examples once have time. cc @jinhyukchang in case he has time to take a look.

Reply:
@feng-tao I really appreciate you opinion, and all that you've done open sourcing this project. 

However, it seems like we disagree on the possibility of this being merged. In the spirit of healthy community and collaboration, I think it would be best to get more than just our opinions on this. Including @jinhyukchang is a good start, but, I'd like to see a few more folks weigh in as well... this probably isn't the first time and certainly won't be the last time that folks won't agree on the implementation of some contributed work. I urge you and others to participate in the process, otherwise we risk squashing creativity IMHO. 

I'd be happy to make modifications as I know my solution does not conform to your expectations as is, but, I'm not going to put the time in until I know there is a chance of it moving forward. 

Perhaps we can ask some other folks on slack? Perhaps some of our other colleagues can weigh in such as @samshuster @Mikhail-Ivanov and Mark Grover (not sure his user name)?

Finally, I hope you enjoy your OOO time 🙇

Reply:
thanks @javamonkey79 for your contribution, yeap, it would be good to ask others opinions on this.

Reply:
Hi @javamonkey79,

First of all, thanks for your contribution. I just took a look at the PR, and I think we have some discrepancy on the understanding and/or the use case of Databuilder jobs.
As you already know, constructing Databuilder job is easy once you have job_config configured and job_config is the one mostly provides flexibility on how we are building Lego blocks. I think the Factory in this PR mostly assumes that job_config is in common pattern which is hard to expect in production use case. I think we can assume this use case for sample loader case, but I am still not sure about the improvement that this will make on sample loader case as well -- please help me understand better :)

Again, thanks for the contribution and let's discuss more!


FYI, here's couple job_configs that we run.

    # job config
    job_config = ConfigFactory.from_dict({
        'extractor.hive_table_last_updated.{}'.format('partitioned_table_where_clause_suffix'):
            partitioned_table_where_clause,
        'extractor.hive_table_last_updated.{}'.format('non_partitioned_table_where_clause_suffix'):
            non_partitioned_table_where_clause,
        'extractor.hive_table_last_updated.{}'.format('file_check_threshold'): 1000,
        'extractor.hive_table_last_updated.extractor.sqlalchemy.{}'.format(
            SQLAlchemyExtractor.CONN_STRING): connection_string(),
        'extractor.hive_table_last_updated.extractor.fs_worker_pool_size': pool_size,
        'extractor.hive_table_last_updated.filesystem.{}'.format(
            FileSystem.DASK_FILE_SYSTEM): s3fs.S3FileSystem(anon=False,
                                                            config_kwargs={'max_pool_connections': pool_size}),
        'loader.filesystem_csv_neo4j.{}'.format(FsNeo4jCSVLoader.NODE_DIR_PATH):
            node_files_folder,
        'loader.filesystem_csv_neo4j.{}'.format(FsNeo4jCSVLoader.RELATION_DIR_PATH):
            relationship_files_folder,
        'loader.filesystem_csv_neo4j.{}'.format(FsNeo4jCSVLoader.SHOULD_DELETE_CREATED_DIR):
            True,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NODE_FILES_DIR):
            node_files_folder,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.RELATION_FILES_DIR):
            relationship_files_folder,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_END_POINT_KEY):
            neo4j_endpoint,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_USER):
            neo4j_user,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_PASSWORD):
            neo4j_password,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.JOB_PUBLISH_TAG):
            kwargs['templates_dict'].get(neo4j_csv_publisher.JOB_PUBLISH_TAG)
    })


    job_config = ConfigFactory.from_dict({
        'extractor.table_column_usage_aggregate.extractor.dbapi.{}'.format(DBAPIExtractor.CONNECTION_CONFIG_KEY):
            hive_conn,
        'extractor.table_column_usage_aggregate.extractor.dbapi.{}'.format(DBAPIExtractor.SQL_CONFIG_KEY): hql,
        'extractor.table_column_usage_aggregate.extractor.dbapi.model_class':
            'databuilder.models.presto_query_logs.PrestoQueryLogs',
        'extractor.table_column_usage_aggregate.{}'.format(RAW_EXTRACTOR): db_api_extractor,
        'extractor.table_column_usage_aggregate.transformer.regex_str_replace.{}'.format(ATTRIBUTE_NAME):
            'query_text',
        'extractor.table_column_usage_aggregate.transformer.regex_str_replace.{}'.format(REGEX_REPLACE_TUPLE_LIST):
            [('\\n', '\n'), ('\\', ' '), ('"', '')],
        'extractor.table_column_usage_aggregate.transformer.sql_to_tbl_col_usage.{}'
            .format(SqlToTblColUsageTransformer.COLUMN_EXTRACTION_TIMEOUT_SEC): 600,
        'extractor.table_column_usage_aggregate.transformer.sql_to_tbl_col_usage.{}'
            .format(SqlToTblColUsageTransformer.DATABASE_NAME): 'hive',
        'extractor.table_column_usage_aggregate.transformer.sql_to_tbl_col_usage.{}'
            .format(SqlToTblColUsageTransformer.SQL_STATEMENT_ATTRIBUTE_NAME): 'query_text',
        'extractor.table_column_usage_aggregate.transformer.sql_to_tbl_col_usage.{}'
            .format(SqlToTblColUsageTransformer.USER_EMAIL_ATTRIBUTE_NAME): 'user',
        'extractor.table_column_usage_aggregate.transformer.sql_to_tbl_col_usage.extractor.hive_table_metadata.{}'
            .format(HiveTableMetadataExtractor.WHERE_CLAUSE_SUFFIX_KEY): where_clause_suffix,
        'extractor.table_column_usage_aggregate.transformer.sql_to_tbl_col_usage.extractor.'
        'hive_table_metadata.extractor.sqlalchemy.{}'.format(SQLAlchemyExtractor.CONN_STRING): connection_string(),
        'loader.filesystem_csv_neo4j.{}'.format(FsNeo4jCSVLoader.NODE_DIR_PATH):
            node_files_folder,
        'loader.filesystem_csv_neo4j.{}'.format(FsNeo4jCSVLoader.RELATION_DIR_PATH):
            relationship_files_folder,
        'loader.filesystem_csv_neo4j.{}'.format(FsNeo4jCSVLoader.SHOULD_DELETE_CREATED_DIR):
            True,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NODE_FILES_DIR):
            node_files_folder,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.RELATION_FILES_DIR):
            relationship_files_folder,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_END_POINT_KEY):
            neo4j_endpoint,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_USER):
            neo4j_user,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_PASSWORD):
            neo4j_password,
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_CREATE_ONLY_NODES):
            [User.USER_NODE_LABEL],
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.JOB_PUBLISH_TAG):
            kwargs['templates_dict'].get(neo4j_csv_publisher.JOB_PUBLISH_TAG)
    })

Reply:
@jinhyukchang thanks for the explanation, and showing me this example. I am beginning to see where our differences lie. 

Fundamentally, we cannot do extraction in our dags, because databricks does not expose the hive metastore fully at this point. So, we fall back to CSV. So actually, the sample data loader is exactly what we use now! :)

Keeping this in mind, when I look at your example, and compare it to the factory code, I notice that the loader and publisher code is actually duplicated, which is what the factory aims to solve. 

I would propose:
* I extract out a parent job factory class that handles this duplication
* I keep the existing csv loader based factory as it can still be useful for csv loading specifically 
* I add job config to the constructor arg, such that it takes precedence over default

wdyt? Does it make sense?